A53-xds-ignore-resource-deletion.md

```diff
@@ -0,0 +1,92 @@
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+This design specifies an option for the gRPC XdsClient that allows it to
+ignore resource deletions sent by the xDS server, thus making it more
+fault tolerant in the face of control plane bugs or misconfigurations
+that delete required resources out from under gRPC.
+
+## Background
+
+Currently, when an xDS server tells a client that an LDS or CDS resource
+has been deleted, gRPC will unconfigure the behavior that had been
+configured by that resource.  For example, on the gRPC client, an LDS
+resource deletion will put the channel in `TRANSIENT_FAILURE`, and a CDS
+resource deletion will cause the channel to fail RPCs bound for that
+cluster.  On the gRPC server, an LDS resource deletion will cause the
+server to stop accepting incoming connections, effectively stopping
+the server.
+
+This behavior means that gRPC will not behave in a fault-tolerant manner
+in the face a control plane bug or misconfiguration that deletes a
+resource that it was actually using, which can lead to an outage.  To avoid
+this, we would like gRPC to ignore the deletion of any required resource.
+
+However, if the control plane does not have a mechanism to alert human
+operators when a client is requesting a resource that does not exist,
+then it is not necessarily safe for the client to simply ignore these
+deletions.  This could lead to a situation where the control plane
+deploys a bad update, but the operators think everything is fine because
+the clients are still working, and then a problem is discovered days or
+even months later when the clients are restarted and fail to come back up.
+This could lead to a different type of outage than the one that this design
+is intended to address.
+
+Because we cannot know whether a given control plane has appropriate
+alerting for human operators when a client is requesting a resource that
+does not exist, we propose to make this behavior configurable in gRPC.
+
+### Related Proposals: 
+
+* [gRFC A27: xDS-Based Global Load Balancing](A27-xds-global-load-balancing.md)
+* [A30: xDS v3 Support](A30-xds-v3.md)
+* [A47: xDS Federation](A47-xds-federation.md)
+
+## Proposal
+
+The ability to ignore resource deletions from the xDS server will be
+configured using the server feature mechanism introduced in [gRFC
+A30](A30-xds-v3.md).  Note that, as per [gRFC A47](A47-xds-federation.md),
+a separate server feature list can be configured for each server in a
+federated configuration, so this feature can be enabled on a per-server
+basis.
+
+The new server feature will be the string `ignore_resource_deletion`; if
+present, this feature will be enabled.
+
+When this feature is enabled, the XdsClient will not invoke the
+watchers' `OnResourceDoesNotExist()` method when a resource is deleted,
+nor will it remove the existing resource value from its cache.
+
+Currently, this behavior will affect only LDS and CDS, because those are
```

Comment:
Good point.  Updated the text accordingly.

Reply:
Actually... on second thought, I'm not sure about this.  In C-core, we currently invoke the watcher's `OnError()` method when the resource timer fires, not the `OnResourceDoesNotExist()` method.  That's arguably wrong if this is the initial ADS stream after the client started up and it does not yet have a cached copy of the resource.  But I'm not sure that `OnResourceDoesNotExist()` is actually correct in the case where we were sent the resource on the old ADS stream but the timeout fired after the stream was closed and reestablished.  I think we probably shouldn't actually start the timer at all in that case, especially since it's possible that the server might not resend the resource as an optimization (see [spec](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-resource-type-instance-version)).

I need to think about this a bit more.

Reply:
So, having thought about this overnight, I believe that we should not actually start the timer in the first place upon ADS stream reconnects, because (a) starting the timer would make it impossible for servers to avoid resending the resource based on the `version_info` field, and (b) it doesn't really buy us anything to start the timer, because even if we could reliably tell that the resource did not exist after a stream restart, we cannot do so if the stream does not restart, so this still isn't really a reliable mechanism.  To be fair, the optimization based on the `version_info` field is not necessarily all that useful in the general case (the main case where it's useful is LDS/CDS wildcard subscriptions, which gRPC does not use), but there are some corner cases where it may come in handy, and the spec does allow it, so I think we need to handle this correctly for interoperability reasons.

So, given that, I think this gRFC does actually apply only for LDS and CDS in the SotW protocol variant, so I've restored the original wording here.

Reply:
I've sent grpc/grpc#29668 to fix this behavior in C-core.